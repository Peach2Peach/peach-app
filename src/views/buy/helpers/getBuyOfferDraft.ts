type GetBuyOfferDraftParams = {
  minBuyAmount: number
  maxBuyAmount: number
  meansOfPayment?: MeansOfPayment
}
export const getBuyOfferDraft = ({
  minBuyAmount,
  maxBuyAmount,
  meansOfPayment = {},
}: GetBuyOfferDraftParams): BuyOfferDraft => ({
  type: 'bid',
  creationDate: new Date(),
  lastModified: new Date(),
  amount: [minBuyAmount, maxBuyAmount],
  meansOfPayment,
  paymentData: {},
  releaseAddress: '',
  originalPaymentData: [],
})
