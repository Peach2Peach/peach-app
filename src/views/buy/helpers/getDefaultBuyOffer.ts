type DefaultOfferProps = {
  minBuyAmount: number
  maxBuyAmount: number
  meansOfPayment?: MeansOfPayment
  kyc?: boolean
}
export const getDefaultBuyOffer = ({
  minBuyAmount,
  maxBuyAmount,
  meansOfPayment = {},
  kyc = false,
}: DefaultOfferProps): BuyOfferDraft => ({
  type: 'bid',
  creationDate: new Date(),
  lastModified: new Date(),
  amount: [minBuyAmount, maxBuyAmount],
  meansOfPayment,
  paymentData: {},
  releaseAddress: '',
  originalPaymentData: [],
  kyc,
})
