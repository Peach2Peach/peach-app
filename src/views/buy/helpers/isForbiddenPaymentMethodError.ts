import { isPaymentMethod } from '../../../utils/validation/isPaymentMethod'

export const isForbiddenPaymentMethodError = (
  errorMessage: string | null,
  errorDetails: unknown,
): errorDetails is PaymentMethod[] =>
  errorMessage === 'FORBIDDEN' && Array.isArray(errorDetails) && errorDetails.every(isPaymentMethod)
