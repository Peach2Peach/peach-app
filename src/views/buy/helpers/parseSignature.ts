const signaturePattern = /[a-zA-Z0-9/=+]{88}/u

export const parseSignature = (signature: string) => signature.match(signaturePattern)?.pop() || signature
