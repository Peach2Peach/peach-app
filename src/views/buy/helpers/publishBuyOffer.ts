import { publishPGPPublicKey } from '../../../init/publishPGPPublicKey'
import { saveOffer } from '../../../utils/offer'
import { postBuyOffer } from '../../../utils/peachAPI'

export const publishBuyOffer = async (offerDraft: BuyOfferDraft) => {
  let [result, err] = await postBuyOffer(offerDraft)

  if (err?.error === 'PGP_MISSING') {
    await publishPGPPublicKey()
    const response = await postBuyOffer(offerDraft)
    result = response[0]
    err = response[1]
  }
  if (result) {
    saveOffer({ ...offerDraft, ...result })
    return { offerId: result.id, isOfferPublished: true, errorMessage: null }
  }
  return { isOfferPublished: false, errorMessage: err?.error || 'POST_OFFER_ERROR' }
}
