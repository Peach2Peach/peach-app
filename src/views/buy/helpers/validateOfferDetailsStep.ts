import { getPaymentData, getSelectedPaymentDataIds } from '../../../utils/account'
import { isDefined } from '../../../utils/array/isDefined'
import { hasMopsConfigured } from '../../../utils/offer'
import { isValidPaymentData } from '../../../utils/paymentMethod'

export const validateOfferDetailsStep = (
  offerDraft: BuyOfferDraft,
  preferredPaymentMethods: Settings['preferredPaymentMethods'],
) =>
  !!offerDraft.amount
  && hasMopsConfigured(offerDraft)
  && getSelectedPaymentDataIds(preferredPaymentMethods).map(getPaymentData)
    .filter(isDefined)
    .every(isValidPaymentData)
