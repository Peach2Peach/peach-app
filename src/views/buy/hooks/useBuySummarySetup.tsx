import React, { useEffect, useState } from 'react'
import shallow from 'zustand/shallow'
import { WalletIcon } from '../../../components/icons'
import { useHeaderSetup, useNavigation } from '../../../hooks'
import { useShowErrorBanner } from '../../../hooks/useShowErrorBanner'
import pgp from '../../../init/pgp'
import { useSettingsStore } from '../../../store/settingsStore'
import { account, getMessageToSignForAddress, updateTradingLimit } from '../../../utils/account'
import i18n from '../../../utils/i18n'
import { saveOffer } from '../../../utils/offer'
import { getTradingLimit, postBuyOffer } from '../../../utils/peachAPI'
import { isValidBitcoinSignature } from '../../../utils/validation'
import { peachWallet } from '../../../utils/wallet/setWallet'

const getAndUpdateTradingLimit = () =>
  getTradingLimit({}).then(([tradingLimit]) => {
    if (tradingLimit) {
      updateTradingLimit(tradingLimit)
    }
  })

export const useBuySummarySetup = () => {
  const navigation = useNavigation()
  const showErrorBanner = useShowErrorBanner()

  const [peachWalletActive, payoutAddress, payoutAddressLabel, payoutAddressSignature] = useSettingsStore(
    (state) => [state.peachWalletActive, state.payoutAddress, state.payoutAddressLabel, state.payoutAddressSignature],
    shallow,
  )
  const [releaseAddress, setReleaseAddress] = useState('')
  const [message, setMessage] = useState('')
  const [canPublish, setCanPublish] = useState(false)
  const [isPublishing, setIsPublishing] = useState(false)
  const [messageSignature, setMessageSignature] = useState(
    !peachWalletActive && payoutAddressSignature ? payoutAddressSignature : '',
  )
  const walletLabel = peachWalletActive ? i18n('peachWallet') : payoutAddressLabel

  const goToSetupPayoutWallet = () =>
    payoutAddress ? navigation.navigate('signMessage') : navigation.navigate('payoutAddress', { type: 'payout' })

  const publishOffer = async (offer: BuyOfferDraft) => {
    setIsPublishing(true)

    await pgp() // make sure pgp has been sent
    const [result, err] = await postBuyOffer(offer)

    if (result) {
      getAndUpdateTradingLimit()
      saveOffer({ ...offer, id: result.offerId } as BuyOffer)
      navigation.replace('offerPublished', { isSellOffer: false })
      return
    }

    setIsPublishing(false)
    showErrorBanner(err?.error || 'POST_OFFER_ERROR')
  }

  useHeaderSetup({
    title: i18n('buy.summary.title'),
    icons: [{ iconComponent: <WalletIcon />, onPress: () => navigation.navigate('selectWallet', { type: 'payout' }) }],
  })

  useEffect(() => {
    setCanPublish(isValidBitcoinSignature(message, releaseAddress, messageSignature))
  }, [releaseAddress, message, messageSignature])

  useEffect(() => {
    ;(async () => {
      const address = peachWalletActive ? await peachWallet.getReceivingAddress() : payoutAddress
      if (!address) return

      const messageToSign = getMessageToSignForAddress(account.publicKey, address)
      setReleaseAddress(address)
      setMessage(messageToSign)
      setMessageSignature(
        peachWalletActive ? peachWallet.signMessage(messageToSign, address) : payoutAddressSignature || '',
      )
    })()
  }, [payoutAddress, payoutAddressSignature, peachWalletActive])

  return {
    releaseAddress,
    walletLabel,
    message,
    messageSignature,
    canPublish,
    publishOffer,
    isPublishing,
    goToSetupPayoutWallet,
  }
}
