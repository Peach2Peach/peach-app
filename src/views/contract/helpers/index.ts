export { isPSBTForBatch } from "./isPSBTForBatch";
export { shouldShowTradeStatusInfo } from "./shouldShowTradeStatusInfo";
export { verifyReleasePSBT as verifyPSBT } from "./verifyReleasePSBT";
