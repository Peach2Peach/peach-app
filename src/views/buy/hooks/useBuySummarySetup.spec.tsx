import { act, renderHook } from '@testing-library/react-native'
import { account1 } from '../../../../tests/unit/data/accountData'
import { NavigationWrapper, replaceMock } from '../../../../tests/unit/helpers/NavigationWrapper'
import { setPaymentMethods } from '../../../paymentMethods'
import { PaymentMethodForbiddenPaypal } from '../../../popups/info/PaymentMethodForbiddenPaypal'
import { useConfigStore } from '../../../store/configStore'
import { usePopupStore } from '../../../store/usePopupStore'
import { setAccount } from '../../../utils/account'
import { PeachWallet } from '../../../utils/wallet/PeachWallet'
import { setPeachWallet } from '../../../utils/wallet/setWallet'
import { useBuySummarySetup } from './useBuySummarySetup'

const publishBuyOfferMock = jest.fn().mockResolvedValue({ offerId: '123', isOfferPublished: true, errorMessage: null })
jest.mock('../helpers/publishBuyOffer', () => ({
  publishBuyOffer: (...args: unknown[]) => publishBuyOfferMock(...args),
}))

jest.mock('../../../utils/validation/isValidBitcoinSignature', () => ({
  isValidBitcoinSignature: jest.fn().mockReturnValue(true),
}))
describe('useBuySummarySetup', () => {
  beforeAll(() => {
    setPaymentMethods([{ id: 'paypal', currencies: ['EUR'], anonymous: false }])
  })
  beforeEach(() => {
    // @ts-expect-error mock doesn't need args
    setPeachWallet(new PeachWallet())
  })

  it('should show offer grouphug announcement if not seen before after publishing', async () => {
    useConfigStore.getState().setHasSeenGroupHugAnnouncement(false)
    setAccount({ ...account1, publicKey: '02d13a5d45bbbf5ef604f01530d22ce3a787c36a78bcb7bd57b2d90fd098686f37' })

    const { result } = renderHook(useBuySummarySetup, { wrapper: NavigationWrapper })
    await act(() => result.current.publishOffer())
    expect(replaceMock).toHaveBeenCalledWith('groupHugAnnouncement', { offerId: '123' })
  })
  it('should show offer published overlay when offer has been published successfully', async () => {
    useConfigStore.getState().setHasSeenGroupHugAnnouncement(true)
    setAccount({ ...account1, publicKey: '02d13a5d45bbbf5ef604f01530d22ce3a787c36a78bcb7bd57b2d90fd098686f37' })

    const { result } = renderHook(useBuySummarySetup, { wrapper: NavigationWrapper })
    await act(() => result.current.publishOffer())
    expect(replaceMock).toHaveBeenCalledWith('offerPublished', { offerId: '123', isSellOffer: false })
  })
  it('should publish the buy offer with the correct parameters', async () => {
    setAccount({ ...account1, publicKey: '02d13a5d45bbbf5ef604f01530d22ce3a787c36a78bcb7bd57b2d90fd098686f37' })

    const { result } = renderHook(useBuySummarySetup, { wrapper: NavigationWrapper })
    await act(() => result.current.publishOffer())
    expect(result.current.isPublishing).toBe(false)
    expect(publishBuyOfferMock).toHaveBeenCalledWith({
      amount: [0, Infinity],
      meansOfPayment: {},
      message: 'I confirm that only I, peach02d13a5d, control the address bcrt1qwype5wug33a6hwz9u2n6vz4lc0kpw0kg4xc8fq',
      messageSignature: 'IH9ZjMHG1af6puAITFTdV5RSYoK1MNmecZdhW0s4soh4EIAz4igtVQTec5yj4H9Iy7sB6qYReRjGpE3b4OoXSLY',
      originalPaymentData: [],
      paymentData: {},
      releaseAddress: 'bcrt1qwype5wug33a6hwz9u2n6vz4lc0kpw0kg4xc8fq',
      type: 'bid',
      walletLabel: 'Peach wallet',
      maxPremium: null,
    })
  })
  it('should open payment method forbidden popup for respective API error', async () => {
    setAccount({ ...account1, publicKey: '02d13a5d45bbbf5ef604f01530d22ce3a787c36a78bcb7bd57b2d90fd098686f37' })

    publishBuyOfferMock.mockResolvedValueOnce({ errorMessage: 'FORBIDDEN', errorDetails: ['paypal'] })
    const { result } = renderHook(useBuySummarySetup, { wrapper: NavigationWrapper })
    await act(() => result.current.publishOffer())
    expect(result.current.isPublishing).toBe(false)
    expect(usePopupStore.getState()).toEqual({
      ...usePopupStore.getState(),
      content: <PaymentMethodForbiddenPaypal />,
    })
  })
})
