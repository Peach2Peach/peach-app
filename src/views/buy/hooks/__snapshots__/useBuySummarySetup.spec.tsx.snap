// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useBuySummarySetup should open payment method forbidden popup for respective API error 1`] = `
<View
  style={
    {
      "borderRadius": 16,
      "marginLeft": 12,
      "marginRight": 12,
      "overflow": "hidden",
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "alignItems": "stretch",
          "backgroundColor": "#FEEDE5",
          "gap": 12,
          "paddingBottom": 24,
          "paddingLeft": 24,
          "paddingRight": 24,
          "paddingTop": 16,
        },
        [
          {
            "backgroundColor": "#D7F2FE",
          },
          {
            "alignItems": "stretch",
          },
        ],
      ]
    }
  >
    <Text
      allowFontScaling={false}
      style={
        [
          {
            "color": "#2B1911",
            "fontFamily": "Baloo2-Regular",
            "fontSize": 16,
            "letterSpacing": 0.16,
            "lineHeight": 24,
          },
          undefined,
          false,
        ]
      }
    >
      You do not have access to Paypal yet. Due to PayPal high risks of charge back, this payment method only becomes accessible when: 

- You unlock the #superTrader badge (more than 20 trades completed)

- Your reputation score is minimum 4.5 peaches

- You already traded at least CHF 2100 during the past yearÂ 
    </Text>
  </View>
  <View
    style={
      [
        {
          "alignItems": "center",
          "alignSelf": "stretch",
          "backgroundColor": "#F56522",
          "flexDirection": "row",
          "justifyContent": "space-between",
        },
        {
          "backgroundColor": "#099DE2",
        },
      ]
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "flexGrow": 1,
          "gap": 4,
          "opacity": 1,
          "paddingBottom": 8,
          "paddingLeft": 24,
          "paddingRight": 24,
          "paddingTop": 8,
        }
      }
    >
      <test-file-stub
        fill="#FFFCFA"
        style={
          [
            {
              "height": 16,
              "width": 16,
            },
            undefined,
          ]
        }
      />
      <Text
        allowFontScaling={false}
        style={
          [
            {
              "color": "#2B1911",
              "fontFamily": "Baloo2-Regular",
              "fontSize": 16,
              "letterSpacing": 0.16,
              "lineHeight": 24,
            },
            [
              {
                "color": "#FFFCFA",
                "fontFamily": "Baloo2-SemiBold",
                "fontSize": 16,
                "letterSpacing": 0.16,
                "lineHeight": 28,
              },
              undefined,
            ],
            false,
          ]
        }
      >
        help
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "flexDirection": "row-reverse",
          "flexGrow": 1,
          "gap": 4,
          "opacity": 1,
          "paddingBottom": 8,
          "paddingLeft": 24,
          "paddingRight": 24,
          "paddingTop": 8,
        }
      }
    >
      <test-file-stub
        fill="#FFFCFA"
        style={
          [
            {
              "height": 16,
              "width": 16,
            },
            undefined,
          ]
        }
      />
      <Text
        allowFontScaling={false}
        style={
          [
            {
              "color": "#2B1911",
              "fontFamily": "Baloo2-Regular",
              "fontSize": 16,
              "letterSpacing": 0.16,
              "lineHeight": 24,
            },
            [
              {
                "color": "#FFFCFA",
                "fontFamily": "Baloo2-SemiBold",
                "fontSize": 16,
                "letterSpacing": 0.16,
                "lineHeight": 28,
              },
              undefined,
            ],
            false,
          ]
        }
      >
        close
      </Text>
    </View>
  </View>
</View>
`;
